Implement account registration in the wizard

juliusknorr opened the pull request:
Based on #9 

![2017-07-11-221227_750x507_scrot](https://user-images.githubusercontent.com/3404133/28088501-23f08e28-6686-11e7-914c-61f7a9afd74a.png)

![2017-07-11-221235_750x507_scrot](https://user-images.githubusercontent.com/3404133/28088505-26034f2a-6686-11e7-8d9f-78c931fc3bc4.png)

ToDo:
- [ ] Implement pending account type for registrations
- [ ] Frequently check for registration status
- [ ] Setup HttpCredentials with app password after registration is complete

ckamm committed:
Introduce private link sharing #5023

* SocketAPI has COPL_LOCAL_LINK / EMAIL_LOCAL_LINK commands
* The nautilus and dolphing shell integrations show a submenu from which
  one can share as well as access the private link.
* The SocketAPI provides a new GET_STRINGS command to access localized
  strings.
* The private link can also be accessed from the user/group sharing
  dialog.
* The numeric file id is extracted from the full id to create the
  private link url.

settermjd committed:
Update client docs version string to 2.4.0

Need to update this to match the latest version.

ogoffart committed:
HTTP/2 Support

We need Qt 5.9 for HTTP2 because, even if Qt 5.8 already has support
for it, there is some critical bug in the HTTP2 implementation which
make it unusable [ https://codereview.qt-project.org/186050 and
https://codereview.qt-project.org/186066 ]

When using HTTP2, we can use many more parallel network request, this
is especially good for small file handling

Lower the priority of the GET and PUT propagation jobs, so the quota
or selective sync ui PROPFIND will not be blocked by them

juliusknorr committed:
Add provider list

Signed-off-by: Julius Härtl <jus@bitgrid.net>

Commit:
[tx-robot] updated from transifex

guruz committed:
Update ChangeLog

ogoffart committed:
Don't link dirrectly against openssl

All our crypto code is handled by qt nodaways.
No need to carry this dependency.

Especially since it causes warnings on system where there are
twp openssl version installed:
/usr/bin/ld: warning: libcrypto.so.1.0.0, needed by /usr/lib/libQt5Network.so.5.9.0, may conflict with libcrypto.so.1.1

ogoffart committed:
OAuth: cleanup debug messages

- Add category to the all messages (they did not have it was merged right after
the patch to add category everywhere, but this code did not have it.)

- Make sure there is no warnings in the normal flow. (The wizard does a request
without authentication to determine the auth type)

Commit:
Initialize stop variable.

Avoid stopping the worker thread immediately if the stop variable happens to be true.

Commit:
Initialize stop variable.

Avoid stopping the worker thread immediately if the stop variable happens to be true.

Commit:
shell_integration/windows: Create _newQueries as a non-signaled auto-reset event.

Since ResetEvent() is never called, the call to WaitForMultipleObjects()
can always return immediately once the event has been signaled.

Commit:
shell_integration/windows: Create _newQueries as a non-signaled auto-reset event.

Since ResetEvent() is never called, the call to WaitForMultipleObjects()
can always return immediately once the event has been signaled.

ogoffart committed:
AccountSettings: Triggering log in re-ask about previously rejected certificates #5819

Issue #5819

juliusknorr committed:
Resize list items height and load images from json

Signed-off-by: Julius Härtl <jus@bitgrid.net>

ogoffart committed:
AccessManager: Disable HTTP/2 without TLS

Qt would otherwise still try to do HTTP/2 connection even over "http://".
And that does not work with server that does not support it

Commit:
[tx-robot] updated from transifex

ckamm committed:
SyncEngineTest: Fix date locale related bug

The client is very picky about date strings it accepts. If dates are
formatted with a non-C locale (such as localized weekday names), it
fails to parse it and tests fail in subtle ways.

ckamm committed:
SyncEngineTest: Send 'checksums' in FakePropfind

ckamm committed:
SyncEngineTest: Add network override

This is useful for monitoring what kind of network requests are
sent to the fake server. Such as "did this sync cause an upload?"
and "was there a propfind for this path?". It can also inject
custom replies.

ckamm committed:
Compare the hash of files with identical mtime/size #5589

* For conflicts where mtime and size are identical:

  a) If there's no remote checksum, skip (unchanged)
  b) If there's a remote checksum that's a useful hash, create a
     PropagateDownload job and compute the local hash. If the hashes
     are identical, don't download the file and just update metadata.

* Avoid exposing the existence of checksumTypeId beyond the database
  layer. This makes handling checksums easier in general because they
  can usually be treated as a single blob.

  This change was prompted by the difficulty of producing file_stat_t
  entries uniformly from PROPFINDs and the database.

jturcotte committed:
shell/Windows: Update binaries to build 45

jturcotte committed:
shell/Windows: Update binaries to build 45

ckamm committed:
Link shares: Add "show file listing" option #5837

ckamm committed:
Link share: Remove direct download if unavailable #5837

It would have been much nicer to keep the menu assigned to the
QToolButton, but if one switches away from InstantPopup (to adjust the
entries before they're displayed), the button always gets a menu
indicator that can't be removed.

jturcotte committed:
Remove debug messages by default from sync.engine

Other categories were already set properly.

ckamm committed:
Journal: Don't use a ._ path if it won't work #5633

When synchronizing a folder on a samba share, creating files that begin
with ._ is often forbidden. This prevented the client from creating
its ._sync_abcdef.db file.

Now, it'll check whether the preferred filename is creatable, and if
it isn't it'll use .sync_abcdef.db instead.

The disadvantage is that this alternative path won't be ignored by
older clients - that was the reason for the ._ prefix.

Commit:
[tx-robot] updated from transifex

Commit:
[tx-robot] updated from transifex

Commit:
[tx-robot] updated from transifex

ckamm committed:
Ignore ui: Add the sync journal patterns #5844

The ignore patterns for the journal files are hardcoded. Add them to
the UI to make them discoverable.

Commit:
[tx-robot] updated from transifex

mrow4a committed:
Classify chunked items correctly. Issue #5850

Commit:
[tx-robot] updated from transifex

guruz committed:
owncloudcmd: Explain --debug parameter

guruz committed:
Jenkinsfile: 2.3.3 will ship with Qt 5.

This should fix the current CI run too

guruz committed:
Jenkinsfile: Run full testing

guruz committed:
Jenkinsfile: Run full testing

guruz committed:
Jenkinsfile: Run verbose tests

guruz committed:
Jenkinsfile: Compile in debug because of all the asserts

guruz committed:
Jenkinsfile: Compile in debug because of all the asserts

Commit:
[tx-robot] updated from transifex

ckamm committed:
Merge remote-tracking branch 'origin/2.3'

guruz committed:
test_journal.db: Add checksumtype table

guruz committed:
CSyncSqliteTest: Fix test

guruz committed:
Jenkinsfile: Some comments

Commit:
[tx-robot] updated from transifex

juliusknorr committed:
Filter provider list by free plans or country

Signed-off-by: Julius Härtl <jus@bitgrid.net>

guruz committed:
SyncEngine: Keep local size in SyncItem #5855

Commit:
[tx-robot] updated from transifex

Commit:
[tx-robot] updated from transifex

Commit:
[tx-robot] updated from transifex

Commit:
[tx-robot] updated from transifex

ckamm committed:
Blacklist: Don't let errors become warnings #5516

Before, blacklisted errors were set to FileIgnored status and hence
displayed as warnings. Now, they have their own BlacklistedError
category which allows them to appear as errors in the issues list and in
the shell integration icons.

Commit:
[tx-robot] updated from transifex

ckamm committed:
SyncEngine: Add unittest for SyncFileItem properties #5855

Checks instruction, direction, size, modtime for three common cases.

SamuAlfageme committed:
Update central.owncloud.org link

Desktop client category was renamed, old link was 404 - see: https://central.owncloud.org/t/new-central-categories-you-decide

SamuAlfageme committed:
Update central.owncloud.org link

Desktop client category was renamed, old link was 404 - see: https://central.owncloud.org/t/new-central-categories-you-decide

ckamm committed:
SyncEngine: Introduce overall errors that are not tied to a file #5746

For now we use them for:
* csync errors: This allows them to appear in the sync issues tab
* insufficient local disk space, as a summary of individual file errors

Insufficient remote space will use them too, as might other issues that
are bigger than a single sync item.

juliusknorr committed:
Add registration step to wizard

Commit:
[tx-robot] updated from transifex

Commit:
[tx-robot] updated from transifex

juliusknorr committed:
Make building optional

Signed-off-by: Julius Härtl <jus@bitgrid.net>

juliusknorr committed:
Disable by default

Signed-off-by: Julius Härtl <jus@bitgrid.net>

Commit:
[tx-robot] updated from transifex

Commit:
[tx-robot] updated from transifex

Commit:
[tx-robot] updated from transifex

ckamm committed:
csync tests: Fix for new owncloudcmd return codes

These were introduced in 4af45394f9f61fbc68c9f0a4d2e5387d14052e3b

ckamm committed:
csync tests: remove stray 'print'

guruz committed:
TestSyncJournalDB: Fix concurrent usage on CI

ogoffart committed:
OAuth: Redirects to the server in case of sucessfull login

Requires https://github.com/owncloud/oauth2/pull/45

This commit moves the reply after we got the token reply from
the server, that allows to reply with an error to the browser
if the login does not work.

ogoffart committed:
csync: Remove iconv depedency

We use iconv to normalize the string on mac. But the iconv version
on mac does not support full unicode.
So we will use Qt to normalize the string

ogoffart committed:
Fix check_csync_exclude test

Commit 4697f0274f282cafd28cbad07066efb397dbf21b remove .htaccess from
the sync-exclude.lst, so use another known excluded pattern

ogoffart committed:
Fix check_csync_update

Commit 816096311027bac0f391ea97d0b61b4175e9b8f9 changed the
SQL requests from csync to read from the checksumtype table.
So we need to add this table in the fake DB as well

Commit:
[tx-robot] updated from transifex

ckamm committed:
SyncEngineTest: Fix test reliability

There was a rounding issue in the mtimes which sometimes resulted in an
off-by-one error. Caused by storing a full QDateTime in the FileInfo but
the mtime saved to the disk being truncated to seconds.

ogoffart committed:
main.cpp: Fix a compiler warning

 src/gui/main.cpp:112:9: warning: bool literal returned from 'main' [-Wmain]

Used 1 to keep previous behaviour. I supposed the code was meant
to return success (0), but it does not really matter anyway.

ogoffart committed:
csync: Move the locale<->utf8 conversion to a different file

It's a C++ file so we will be able to use Qt from it

ogoffart committed:
csync: Use Qt for encodeing/decoding filesystem strings

Issues:
 - #5661   On mac, iconv did not support all of unicode and some
           files with emoji in the filename could not be uploaded

 - #5719 , #5676  On linux, we will now support non utf-8 locale

ckamm committed:
AccountState: Add a 1-5min reconnection delay #5872

This only applies when the server was explicitly in maintenance mode or
when it was 503-unavailable.

ckamm committed:
Add a more functional error view #5516 (#5861)

* Add a more functional error view #5516

* Allow filtering of ignores and warnings to see only important bits.
* Navigate from the folder view to the error view by clicking on the
  error list with the red background.
* Move the error list into its own ui file to allow easier extension.
* Fix issue around tab id handling in ActivitySettings.
* Rename "Action" column to "Issue".
* Change mouse cursor to hand over button and new error list area

Several OSX fixes provided by guruz.

guruz committed:
check_vio_ext: Align with new UTF8 behaviour on macOS

jturcotte committed:
SocketApi: Improve logging

jturcotte committed:
SocketApi: Adjust the format of the mailto: URL

Minor tweaks that weren't actually an issue, but just in case.

jturcotte committed:
SocketApi: Add GET_STRINGS:BEGIN and END

The Windows shell extension relied on the response of
SHARE_MENU_TITLE to advance its state machine, but in order
to use the new GET_STRINGS instead, we need to know when the
last string was received. Also add BEGIN for consistency.

jturcotte committed:
shell/windows: Use a submenu to include private link actions

Refactor things a bit to be able to reuse some code and
clean things up.

jturcotte committed:
shell/macos: Use a submenu to include private link actions

jturcotte committed:
shell/macos: Remove legacy (< 10.10) overlay icons

Having to update the plugins to add a submenu in the context menu,
it's more cost-effective to remove support for legacy macOS versions
than implementing and testing this setup.

Commit:
[tx-robot] updated from transifex

Commit:
[tx-robot] updated from transifex

hefee committed:
Use OWNCLOUD_BIN_DIR directly in C++ code.

No need to do a STR replacement and the macro changing the path.
It actually broke test execution for Debian.

See https://bugs.debian.org/844937

ckamm committed:
SocketAPI: Change strings to a consistent _MENU_TITLE postfix

ckamm committed:
shell integrations: Change share menu title to "Share..."

Since the appname is in the context submenu title already.

Also drop APPNAME string.

ckamm committed:
Fix crash on account deletion

Calling forgetSensitiveData() on account deletion leads to a timer for
clearQNAMCache() being queued. Then the Account object is deleted. The
Credentials object stays alive for now because it has a deleteLater
deleter.

If the timer calls into a slot on the Credentials object, the _account
pointer will be invalid at this time.

As a workaround, move the target slot to Account - that way it will not
be called as the account object is already destroyed.

However since Account and Credentials are mutually dependent, it would
be much preferable if their lifetimes were linked, avoiding this
category of bugs.

The current behavior was introduced in
d40c56eda561e3a541bf1b23f70fa8d659d3037e and I currently don't
understand why - maybe there's another way of dealing with the problem
that existed then.

ckamm committed:
FolderMan: Add missing folderListChanged emits

She signal of a folder list change was not emitted when folders were
deleted from the list.

ckamm committed:
IssuesTab: Hide filters if they wouldn't be useful

As suggested by @dragotin

Commit:
[tx-robot] updated from transifex

Commit:
[tx-robot] updated from transifex

ckamm committed:
Account/Credentials: Have identical lifetimes

The QNAM may continue to outlive both.

Rename Credentials::getQNAM() to createQNAM() while we're at it - it's
used to make a new QNAM that will subsequently be owned by the Account
object.

See d01065b9a12e69ca493a232f3a8e8f3d416fed52 for rationale.

Relates to
d40c56eda561e3a541bf1b23f70fa8d659d3037e
147cf798a6f13c9b53a9f1fb2db1ef26c8c63273

ckamm committed:
PropagateUpload: Put upload error handling in one function

ckamm committed:
PropagateUpload: Better messaging for 507 #5537

It now produces a summary error message indicating the problem.

Adjust blacklist database table to contain 'errorCategory'. This is
useful for two things:
  - Reestablishing summary messages based on blacklisted errors. For
    example if we don't retry a 507ed file, we still want to show the
    message about space on the server
  - Selectively wiping the blacklist: When we have ui for something like
    "I deleted some files, please retry all files now!", we want to
    delete all blacklist entries of a specific category only.

jnweiger committed:
Update building.rst

jturcotte committed:
shell/windows Build 46: Private links context menu

Commit:
[tx-robot] updated from transifex

juliusknorr committed:
Use model class for providers

Signed-off-by: Julius Härtl <jus@bitgrid.net>

Commit:
[tx-robot] updated from transifex

Commit:
[tx-robot] updated from transifex

Commit:
[tx-robot] updated from transifex

ckamm committed:
ProgressInfo: Carry a sync status

* A bunch of code was determining sync status by ad-hoc comparing some
  progress info fields. It can now just check the status, making it
  easier to comprehend.
* There's a clear indication for "a new sync is starting", which helps
  wiping the issues tab at the right time.

ckamm committed:
Add ElidedLabel

A label that adjusts its text based on Qt::TextElideMode.

ckamm committed:
IssuesWidget: Hide folder column when possible

ckamm committed:
IssuesWidget: Add button to retry 507 errors #5537

Since these errors are blacklisted, it can take up to 24h to retry items
that had a 507 error for a while. This way users can intervene and cause
an upload attempt immediately.

guruz committed:
ChangeLog: 2.3.3

SamuAlfageme committed:
More refs. to corresponding issue/PR on 2.3.3 changelog

juliusknorr committed:
Make registration request and show final page in wizard

Signed-off-by: Julius Härtl <jus@bitgrid.net>

Commit:
[tx-robot] updated from transifex

ckamm committed:
IssuesWidget: Fix insertion of sync errors before item errors

ckamm committed:
Merge remote-tracking branch 'origin/2.3'

ckamm committed:
SyncEngineTest: _size now correct for uploads #5855

ckamm committed:
Conflicts: Detect and show in issues tab

Incidentally fixes a potential issue where conflicts were silently-
ignored and thus deleted if the parent folder was deleted.

ckamm committed:
AccountSettings: Draw a box to indicate pending conflicts

ckamm committed:
Account: Send a X-Request-ID header #5853

ckamm committed:
Ensure qsrand is called

ogoffart committed:
OAuth2: Have a link to the browser in the owncloud UI

When the browser is open, ad a link in the ui to re-open
the browser.

Issue #5893

kashyap2690 committed:
Broken URL fixed in Dockerfile

Url to mingw repo fixed.

ogoffart committed:
OAuth: Better message when loggin in with the wrong username

Since the user is already in the browser, put the error message in the
browser with a message to log out and then log in as the right user.

Issue #5895

guruz committed:
X-Request-ID: Send per request not per account #5853

kaustubh0x77 committed:
Fix for issue number 5692

guruz committed:
csync_vio: Better log for stat errors

ogoffart committed:
OAuth: Add the user in the authorize call

Issues: #5897, https://github.com/owncloud/oauth2/issues/48

ogoffart committed:
PropagateUpload: Remove the Qt workaround against data corruption

This was only for Qt <= 5.4, which we don't really support anymore.
And even then, the data corruption is happens very seldomly anyway.

kaustubh0x77 committed:
Change error message. Removed Qt5 if check

ogoffart committed:
Merge pull request #5326 from owncloud/cleanup_tests

Use OWNCLOUD_BIN_DIR directly in C++ code.

ogoffart committed:
Merge pull request #5905 from kaustubh-welankar/issue_5692_pull

Fix for issue number #5692 : 

Abort owncloudcmd if the capabilities call fails.

ogoffart committed:
Propagator: fix a qMax which should have been a qMin

ogoffart committed:
OAuth: Fix refresh of token after expiration

Before commit d3b00532b1dd9f44cc606e6738b53345c37582cf,
fetchFromKeychain was called everytime we detect that the creds are
invalid (in AccountState::slotInvalidCredentials)
But since that commit, AccountState was calling askFromUser directly,
breaking the refresh of the token.

So I made sure AccountState::slotInvalidCredentials still calls
refreshAccessToken.

Another change that was made was too be sure to clear the cookies
in HttpCredentials::invalidateToken even when we are only clearing the
access_token. That's because the session with a cookie may stay valid
longer than the access_token

ogoffart committed:
buildint.rst: Put back reference to openssl on windows

OpenSSL might no longer be needed to build the client, but we still need
it to run it.

That reverts part of commit 6e57b0219edd7dcc99509620277bb53478d958a6 for
this file.

Fix #5858
Reported in https://central.owncloud.org/t/locally-built-client-error-creating-ssl-context/8264

ogoffart committed:
Revert "Revert "Discovery: consider also the "shared by me" as shared""

This reverts commit efa7821dd28491ad0b93fec53be6d4ad0f94f19f.

This reverts the revert, but also add a check that the server version
is bigger than 10.0

Issue #4788

ogoffart committed:
AccountState: reset _waitingForNewCredentials when signin in

Just to force the logic to re-ask the credenticals, in case we were
already asking them when singin off.

Issue: https://github.com/owncloud/client/issues/5893#issuecomment-316949686

rullzer commented on src/gui/ocsregistrationjob.h:
```diff
@@ -0,0 +1,47 @@
+     * Result of the OCS request
+     * The value parameter is only set if this was a put request.
+     * e.g. if we set the password to 'foo' the QVariant will hold a QString with 'foo'.
+     * This is needed so we can update the share objects properly
```
Copy paste :wink:

rullzer commented on src/gui/wizard/owncloudproviderlistpage.h:
```diff
@@ -0,0 +1,50 @@
+    void initializePage();
+public slots:
+    void toggleFreePlans(bool state);
+    void setCountry(QString current);
```
use http://doc.qt.io/qt-5/qlocale.html#Country-enum ?

rullzer reviewed (changes requested):
I'll have a more detailed look soon

jturcotte committed:
Don't keep the list of touched files for the whole sync

We only want to know if they were touched within the last 15 seconds,
so change the data structure to use a QMultiMap, and sort them by
QElapsedTimer. This allows us to iterate over old entries ordered by
time and to stop once we find a recent entry.

This makes the look-up slower but in most cases the folder watcher
will report any change within milliseconds, and we start from the
most recent. What this really makes slower are actual user file
changes while a fast sync is underways which will need to iterate
over the whole map to find out the file isn't there.

This reduces the growth of the memory usage when downloading a large
amount of files.

ogoffart committed:
OAuth: clear refresh token when the server claim not to support oauth

Allow upgrade path when the server removes support for oauth
Relates: https://github.com/owncloud/client/issues/5848#issuecomment-317353049

We also need to force the account to commit the config to the disk,
otherwise we may not register we are no longer using owncloud and we
risk sending the password as the token to the token refresh API call

jturcotte committed:
windows: Fix a memory leak in FileSystem::longWinPath

Also use c_path_to_UNC directly instead of doing an extra
UTF-8->wchar conversion.

guruz committed:
encoding_tests: No more iconv

jturcotte committed:
Fix the selective sync notification folder list being cropped

The maximumHeight would stay at the last animated value.

Issue #5492

jturcotte committed:
General settings: Put all advanced checkboxes in a layout

The mac style has the concept of Layout Item Rectangle that bleeds
widgets margins into parent margins. This unfortunately doesn't work
when the parent layout doesn't have any margin, like we do when we
have the hierarchy:
 Vertical(normal margin)
   / Horizontal(no margin, uses parent layout spacing)
     /Widget

I guess Qt should propagate remaining uneaten margins to grand-parent
layouts and so on to have this work properly, but nobody seems to
have touched that code since Qt 4.4.
So just try to make the problem less worse by making sure that
all checkboxes we want to align are either in the same layout, or
in a loyout of their own.

Issue #5492

ogoffart committed:
FolderWizard: fix crash in FolderWizardRemotePath::slotHandleLsColNetworkError

The Job is a LsColJob, not a MkColJob!

Reproduce by entering a name with invalid character that cause an error 400
in the folder wizard's remote path line edit.

(Relates issue #5924)

jturcotte committed:
macos: Add a sidebar icon

In the process, use an iconset to generate the icns using iconutil.
Also add some missing icon resolutions according to the guidelines.

Issue #296

settermjd committed:
Clean up the client docs removing old, outdated, content

This makes changes as suggested by @michaelstingl in
https://github.com/owncloud/documentation/issues/3240.

settermjd committed:
Merge pull request #5929 from owncloud/doc-cleanup

Clean up the client docs removing old, outdated, content

settermjd committed:
Merge pull request #5762 from owncloud/update-docs-to-2.3.2

Update client docs version string to 2.3.2

guruz committed:
Sync Folder: Add file manager favorite also in folder wizard #455

Not only in the account wizard.

guruz committed:
ChangeLog: More 2.4.0 stuff

guruz committed:
Upload: Log the checksum / file hash

rullzer committed:
Add compilestep CI

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer committed:
Merge pull request #11 from nextcloud/drone

Add compilestep CI

juliusknorr committed:
Use proper defaults for showing provider list

Signed-off-by: Julius Härtl <jus@bitgrid.net>

juliusknorr committed:
Unify button layout in provider list

Signed-off-by: Julius Härtl <jus@bitgrid.net>

rullzer committed:
Merge branch 'master' into upstream_master

juliusknorr committed:
Merge pull request #13 from nextcloud/upstream_master

Upstream master

juliusknorr committed:
Fix styling issues

Signed-off-by: Julius Härtl <jus@bitgrid.net>

juliusknorr commented:
Closing in favour of https://github.com/nextcloud/gsoc_client/pull/17
